### PR TITLE
Make the state parameter verification optional

### PR DESCRIPTION
**Pull request by skycocker**
Addresses https://github.com/omniauth/omniauth_openid_connect/issues/122.

### CODE REVIEW

File: lib/omniauth/strategies/openid_connect.rb [hunk 1]
```diff
@@ -107,7 +108,7 @@ def request_phase
       def callback_phase
         error = params['error_reason'] || params['error']
         error_description = params['error_description'] || params['error_reason']
-        invalid_state = params['state'].to_s.empty? || params['state'] != stored_state
+        invalid_state = options.verify_state && (params['state'].to_s.empty? || params['state'] != stored_state)
```
Comment (syakovyn) [inline lib/omniauth/strategies/openid_connect.rb:111]: I'd not exclude state mismatch from the check. Only a case when there is no state stored.

Comment (skycocker) [reply lib/omniauth/strategies/openid_connect.rb]: Fair - made the changes.

**Review by syakovyn — APPROVED** [review]

**Comment by stanhu** [comment]
Can you rebase `master`?

**Comment by stanhu** [comment]
There are also Rubocop failures in https://github.com/omniauth/omniauth_openid_connect/actions/runs/3338515447/jobs/5862634541.

**Comment by skycocker** [comment]
Addressed.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,7 @@
+# v0.4.1 (06.02.2022)
```
Comment (stanhu) [inline CHANGELOG.md:1]: We should probably create an UNRELEASED section, and update this after the official release has been cut.

File: test/lib/omniauth/strategies/openid_connect_test.rb [hunk 3]
```diff
@@ -278,6 +278,65 @@ def test_callback_phase_with_discovery # rubocop:disable Metrics/AbcSize
         strategy.callback_phase
       end
 
+      def test_callback_phase_with_no_state_without_state_verification # rubocop:disable Metrics/AbcSize
+        code = SecureRandom.hex(16)
+        nonce = SecureRandom.hex(16)
```
Comment (stanhu) [inline test/lib/omniauth/strategies/openid_connect_test.rb:283]: This line can be dropped now that `nonce` has been cleaned up in https://github.com/omniauth/omniauth_openid_connect/pull/132.

File: test/lib/omniauth/strategies/openid_connect_test.rb [hunk 4]
```diff
@@ -278,6 +278,65 @@ def test_callback_phase_with_discovery # rubocop:disable Metrics/AbcSize
...
+        access_token.stubs(:id_token).returns(jwt.to_s)
+        client.expects(:access_token!).at_least_once.returns(access_token)
+        access_token.expects(:userinfo!).returns(user_info)
+
+        strategy.call!('rack.session' => { 'omniauth.nonce' => nonce })
+        strategy.callback_phase
+      end
+
+      def test_callback_phase_with_invalid_state_without_state_verification
+        code = SecureRandom.hex(16)
```
Comment (stanhu) [inline test/lib/omniauth/strategies/openid_connect_test.rb:326]: `code` and `nonce` can probably be dropped here due to https://github.com/omniauth/omniauth_openid_connect/pull/132. I think we should consider also moving `state` into the `strategy_test_case.rb`.

**Review by syakovyn — APPROVED** [review]

**Pull request merged**